docs/src/coding-guidelines.md

```diff
@@ -0,0 +1,130 @@
+```c++
+object.callFunction(/*arg0=*/nullptr);
+```
+
+Every function, class, or non-trivial piece of logic should have a comment. Avoid redundant comments for self-explanatory code, but never leave complex code unexplained. Example of redundant comment:
+```c++
+// Increment the counter by 1.  // Redundant, avoid.
+counter++;
+```
+
+Ensure comments are accurate and reflect the current state of the code. Outdated or misleading comments can be worse than no comments at all.
+
+# Code Denesting (Inversion)
+
+Strive to minimize unnecessary indentation without compromising code clarity. One effective way to achieve this is by using early exits and the continue keyword in long loops.
+
+Consider following example:
+```c++
+// ...
+if (modified && isDpsOp(op))
+{
+    // ... some long code ...
+}
+```
+
+It is strongly recommended to format the code as follows:
+```c++
+// ...
+// We just need to transform the op if it was modified
+if (!modified) return;
+
+// We also just need to transform the op if it is DPS op
+if (!isDpsOp(op)) return;
+
+// .. some long code ...
+```
+
+This reduces loop nesting, makes the reasoning behind the conditions clearer, and signals to the reader that there is no subsequent else to worry about, reducing cognitive load. This can significantly improve code readability and comprehension.
+
+# Using namespaces
+
+Namespaces are an important part of C++ programming, providing a way to organize code and avoid naming conflicts. Choose namespace names that reflect the purpose or functionality of the code contained within.
+
+Follow these guidelines when defining namespaces:
+    - Use *lower-case* letters for short, single-word names or those with a clear acronym (e.g., ttnn, mlir).
+    - Use nested namespaces to group logically related code, avoiding too deep or unnecessarily complex hierarchy
+    - Avoid placing code in the global namespace to reduce the potential for name conflicts and ambiguity. Always use specific namespaces. If necessary to use something from the global namespace (such as std), use an explicit std:: prefix rather than importing everything using using namespace std;
+    - Never use using namespace in header files, as it can lead to namespace pollution and potential conflicts. If you must use it, do so in .cpp (source) files where the scope is limited. Instead of importing entire namespaces, use specific imports to avoid ambiguity.
+```c++
+using std::vector;
+using std::string;
+```
+    - When namespaces are long or difficult to type repeatedly, use aliases for ease of use, but keep them meaningful.
+```c++
+namespace utils = mlir::tt::ttnn::utils;
+```
+    - Try to avoid mixing concepts from different namespaces in a single function or class. If a function belongs to one namespace but calls classes from others, ensure the relationships are clear.
```

Comment:
I think we should borrow some of the anti-patterns from google style guide: https://google.github.io/styleguide/cppguide.html#Namespaces

The two main offenders I've seen historically:
> - You may not use a using-directive to make all names from a namespace available.
> - Do not use Namespace aliases at namespace scope in header files except in explicitly marked internal-only namespaces, because anything imported into a namespace in a header file becomes part of the public API exported by that file.